sw/device/silicon_creator/lib/shutdown_unittest.cc

```diff
@@ -0,0 +1,399 @@
+#include "otp_ctrl_regs.h"
+
+// FIXME: I can't get ARRAYSIZE from `memory.h` because the definitions of
+// memcpy and friends there conflict with the standard definitions.
+#define ARRAYSIZE(x) ((sizeof x) / (sizeof x[0]))
+
+namespace shutdown_unittest {
+namespace {
+extern "C" {
+// Dummy out base_printf.
+int base_printf(const char *fmt, ...) { return 0; }
+}  // extern "C"
+
+namespace internal {
+// Create a mock for reading OTP words.
+class MockOtp {
+ public:
+  MOCK_METHOD(uint32_t, otp_read32, (uint32_t));
+  virtual ~MockOtp() {}
+};
+
+// Create a mock for shutdown functions.
+class MockShutdown {
+ public:
+  MOCK_METHOD(void, shutdown_software_escalate, ());
+  MOCK_METHOD(void, shutdown_keymgr_kill, ());
+  MOCK_METHOD(void, shutdown_sram_kill, ());
+  MOCK_METHOD(void, shutdown_flash_kill, ());
+  MOCK_METHOD(void, shutdown_epmp_kill, ());
+  MOCK_METHOD(void, shutdown_hang, ());
+  virtual ~MockShutdown() {}
+};
+
+}  // namespace internal
+// Use NiceMock because we aren't interested in the specifics of OTP reads,
+// but we want to mock out calls to otp_read32.
+using MockOtp = mask_rom_test::GlobalMock<testing::NiceMock<internal::MockOtp>>;
+using MockShutdown =
+    mask_rom_test::GlobalMock<testing::StrictMock<internal::MockShutdown>>;
+extern "C" {
+uint32_t otp_read32(uint32_t address) {
+  return MockOtp::Instance().otp_read32(address);
+}
+
+void shutdown_software_escalate(void) {
+  return MockShutdown::Instance().shutdown_software_escalate();
+}
+void shutdown_keymgr_kill(void) {
+  return MockShutdown::Instance().shutdown_keymgr_kill();
+}
+void shutdown_sram_kill(void) {
+  return MockShutdown::Instance().shutdown_sram_kill();
+}
+void shutdown_flash_kill(void) {
+  return MockShutdown::Instance().shutdown_flash_kill();
+}
+void shutdown_epmp_kill(void) {
+  return MockShutdown::Instance().shutdown_epmp_kill();
+}
+void shutdown_hang(void) { return MockShutdown::Instance().shutdown_hang(); }
+}  // extern "C"
+
+using testing::ElementsAre;
+using testing::Invoke;
+using testing::Return;
+using testing::Test;
+
+
+#define PACK(a, b, c, d)                                              \
+  (((uint8_t)(a) << 0) | ((uint8_t)(b) << 8) | ((uint8_t)(c) << 16) | \
+   ((uint8_t)(d) << 24))
+
+#define PASTE(a, b) (a##b)
+#define FULL(name, prod, prodend, dev, rma)                      \
+  {                                                              \
+    name, PASTE(kAlertClass, prod), PASTE(kAlertClass, prodend), \
+        PASTE(kAlertClass, dev), PASTE(kAlertClass, rma)         \
+  }
+
+#define CLASSIFY(name, prod, prodend, dev, rma)               \
+  PACK(PASTE(kAlertClass, prod), PASTE(kAlertClass, prodend), \
+       PASTE(kAlertClass, dev), PASTE(kAlertClass, rma))
+
+// This alert configuration is describe in the Mask ROM Shutdown specification:
+// https://docs.google.com/document/d/1V8hRvQnJhsvddieJbRHS3azbPZvoBWxfxPZV_0YA1QU/edit#
+
+// clang-format off
+#define ALERTS(Xmacro) \
```

Comment:
I prefer not to - to my sense of style, those braces should remain at the initializer site.  The reason (although not used in this file) is so that one could customize the initializer:

```
const Foo foo[] = {
  ALERTS(FOOMACRO),
  "some extra stuff",
  ...
}
```

Reply:
Ok, fine.